Use overloads to make the return types of st.feedback more precise. (with type tests)

wyattscarpenter opened the pull request:
## Describe your changes

Inspired by the adventures of @Asaurus1 in issue #9048, I thought it would be nice for the return type of st.feedback to be narrowed so that my typechecker knows if the function returns 0|1|None or 0|1|2|3|4|None. So, I used `@overload` and `Literal` to achieve this.

## Testing Plan

@Asaurus1, in his pr, is introducing a folder of test code that mypy now gets run over to verify that the return types are correct. This seemed like a good idea to me, so I used the same path and concept.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

wyattscarpenter committed:
Use overloads to make the return types of st.feedback more precise.

Also, add tests for the typing, inspired by @Asaurus1 in issue #9048.

Asaurus1 commented on scripts/mypy:
I see... Now it's a race to who gets to merge conflict first. Challenge accepted 😄

wyattscarpenter replied:
I made sure to add (I hope) the same string, in the same location, so hopeful git will reconcile these seamlessly!

Asaurus1 commented on lib/streamlit/elements/widgets/button_group.py:
```diff
@@ -80,6 +81,8 @@
 
+_FeedbackOptions2: TypeAlias = Literal["thumbs"]
+_FeedbackOptions5: TypeAlias = Literal["faces", "stars"]
 _FeedbackOptions: TypeAlias = Literal["thumbs", "faces", "stars"]
```
Could make this a union of the 2 and 5 options.

wyattscarpenter replied:
I tried this, because it also seemed like a logical idea to me, but it seems like mypy refuses to accept it. Possibly because the mypy version this project uses is not the latest version? (I've had a devil of a time with the dependency system on this project 🤷‍♀️ so I didn't chase the possibility of upgrading). Or maybe I just didn't try the right variant.

raethlein replied:
I have just tested it locally and `_FeedbackOptions: TypeAlias = _FeedbackOptions2 | _FeedbackOptions5` does not work for me either, but `_FeedbackOptions: TypeAlias = Union[_FeedbackOptions2, _FeedbackOptions5]` does

Asaurus1 commented on lib/streamlit/elements/widgets/button_group.py:
```diff
@@ -168,6 +171,28 @@ def _build_proto(
 
 
 class ButtonGroupMixin:
+    @overload  # These overloads are not documented in the docstring, at least not at this time, on the theory that most people won't know what it means. And the Literals here are a subclass of int anyway.
```
This won't pass the linter checks, I'm pretty sure. Make sure you run `make pyformat`.

wyattscarpenter replied:
I was surprised as well, but it ruff tolerates it. It passes my copy of the CI, which I've enabled on my GitHub repo fork. (The true moment of truth still awaits, of course 🍿)

I think this is because comments are a special case that formatting tools know they shouldn't feel free to break up, and the thing before the comment is atomic so it can't be broken up. So ruff check passes it and ruff format doesn't change it.

raethlein commented on lib/streamlit/elements/widgets/button_group.py:
```diff
@@ -168,6 +171,28 @@ def _build_proto(
+    @overload  # These overloads are not documented in the docstring, at least not at this time, on the theory that most people won't know what it means. And the Literals here are a subclass of int anyway.
+    def feedback(
+        self,
+        options: _FeedbackOptions2 = ...,
```
When we use it like this, the autocompletion does not look too nice, because it is not obvious what `_FeedbackOptions2` is referring to. I have just realized that this is already an issue with the existing API where we use `_FeedbackOptions`:
![Screenshot 2024-08-06 at 17 39 18](https://github.com/user-attachments/assets/a3c9a494-618c-4706-a112-06085dbc7042)
vs
![Screenshot 2024-08-06 at 17 39 50](https://github.com/user-attachments/assets/9712d871-d72e-4116-b84f-f20e8b8715fb)

I have just checked and it looks like we have it a little bit mixed in the codebase.
Perhaps we should use the `Literal` types directly for typing the arguments. This means a little bit more duplication, but at least the user-facing API is nicer

wyattscarpenter replied:
Oh, interesting! Yeah, the only reason I kept any concept of a _FeedbackOptions around was because there are two other parts of the code that use it, so not having it be a variable is a DRY violation. But hopefully the typechecker will keep those two other instances from getting out of sync, so maybe it's not so bad. I'm happy to make the change if it improves autocompletion.

This, incidentally, calls into question line [290](https://github.com/streamlit/streamlit/blob/a7720e45a843caf91937fb68004d177c92b41601/lib/streamlit/elements/widgets/button_group.py#L290C8-L290C88): ` if not isinstance(options, list) and options not in get_args(_FeedbackOptions):`. I think the `not isinstance(options, list)` clause is completely unnecessary. If you pass any list in to this function it just, uh... doesn't generate a widget. So I think we should just remove this clause.

raethlein replied:
yeah great catch, thanks! It looks like some left-over from an older state 😞 
With regards to the auto-completion question I'd like to hear the opinion from @snehankekre and @jrieke

wyattscarpenter committed:
replace type alias with literals; also, rm list clause

wyattscarpenter committed:
fix, or try to fix, tests that relied on passing st.feedback wrong type of argument

(old, internal prototype value type, perhaps?)

raethlein commented on lib/tests/streamlit/typing/feedback.py:
```diff
@@ -0,0 +1,26 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
```
I think it could make sense to leave a comment here to explain for what purpose these files exist, something like:

```python
# Files in the typing/ directory are used for catching typing errors as they are checked by our mypy script
```

Otherwise it might be a little bit confusing to have these standalone files, especially they are a now emerging pattern in the library.

wyattscarpenter replied:
Yeah, makes sense. How about a readme.txt in that directory with that verbiage?

raethlein replied:
That sounds like a good idea! probably even better a Readme.md 🙂

wyattscarpenter committed:
Include readme.txt explaining this typing/, using raethlein's proposed verbiage plus my own.

I don't quite recall if mypy will skip the txt on its own or if I will have to change the path to be our homebrew glob

wyattscarpenter committed:
include license header in readme.txt to pass the license header check

feels faintly ridiculous, but I can't deny that this doesn't fall under any other excuse that would warrant an exclusion

wyattscarpenter commented:
Happy to see that other progress in this area had rendered this a merge conflict; I have rebased accordingly.

raethlein commented:
> Happy to see that other progress in this area had rendered this a merge conflict; I have rebased accordingly.

Sorry that this one takes a little bit longer @wyattscarpenter! We appreciate your patience and continued effort 🙂 The other PR is merged so we are getting there 🥳

raethlein commented on lib/tests/streamlit/typing/readme.txt:
nit: README.md might be a nicer format in general, because it is rendered nicely also on GitHub 🙂

raethlein reviewed (approved):
LGTM! Thanks for putting in the effort 🙏 I will let @vdonato or @LukasMasuch make the last pass through here since they have been involved in the other typing PR as well 🙂

wyattscarpenter committed:
Rename readme.txt to readme.md

vdonato reviewed (approved):
LGTM - thanks @wyattscarpenter!
